### PR TITLE
Add benchmark

### PR DESCRIPTION
**Pull request by minborg**
This PR proposed to add a new benchmark outlining potential performance gains with the FFM API in internal JDK classes.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign pull/761/head:pull/761` \
`$ git checkout pull/761`

Update a local copy of the PR: \
`$ git checkout pull/761` \
`$ git pull https://git.openjdk.org/panama-foreign pull/761/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 761`

View PR using the GUI difftool: \
`$ git pr show -t 761`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/761.diff">https://git.openjdk.org/panama-foreign/pull/761.diff</a>

</details>

### CODE REVIEW

**Review by mcimadamore — APPROVED** [review]
Looks good!

File: test/micro/org/openjdk/bench/java/lang/foreign/MemorySegmentVsBits.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    @Benchmark
+    public void bitsEquivalent() {
+        for (int i = 0; i < size; i++) {
+            putLong(bytes, i * Long.BYTES, longs[i]);
+        }
+    }
+
+    @Benchmark
+    public void panamaHeap() {
+        //         for (int i = 0, o = 0; i < size; i++, o += 8) {
```
Comment (mcimadamore) [inline test/micro/org/openjdk/bench/java/lang/foreign/MemorySegmentVsBits.java:100]: Should we drop this line?

**Comment by minborg** [comment]
Here is a run on a MacBook Pro (16-inch, 2021) M1 Max, , macOS 12.6.1:

```
Benchmark                                  (size)  Mode  Cnt    Score   Error  Units
MemorySegmentVsBits.bitsEquivalent              2  avgt   30    3,597 ? 0,021  ns/op
MemorySegmentVsBits.bitsEquivalent             16  avgt   30   20,074 ? 0,082  ns/op
MemorySegmentVsBits.bitsEquivalent             64  avgt   30   81,934 ? 0,328  ns/op
MemorySegmentVsBits.bitsEquivalent            256  avgt   30  322,905 ? 1,041  ns/op
MemorySegmentVsBits.panamaHeap                  2  avgt   30    4,538 ? 0,021  ns/op
MemorySegmentVsBits.panamaHeap                 16  avgt   30    7,040 ? 0,078  ns/op
MemorySegmentVsBits.panamaHeap                 64  avgt   30   16,027 ? 0,022  ns/op
MemorySegmentVsBits.panamaHeap                256  avgt   30   43,778 ? 0,090  ns/op
MemorySegmentVsBits.panamaNative                2  avgt   30    4,564 ? 0,040  ns/op
MemorySegmentVsBits.panamaNative               16  avgt   30    7,036 ? 0,059  ns/op
MemorySegmentVsBits.panamaNative               64  avgt   30   17,237 ? 0,066  ns/op
MemorySegmentVsBits.panamaNative              256  avgt   30   31,617 ? 0,129  ns/op
MemorySegmentVsBits.panamaNativeUnaligned       2  avgt   30    4,334 ? 0,016  ns/op
MemorySegmentVsBits.panamaNativeUnaligned      16  avgt   30    6,944 ? 0,044  ns/op
MemorySegmentVsBits.panamaNativeUnaligned      64  avgt   30   17,220 ? 0,102  ns/op
MemorySegmentVsBits.panamaNativeUnaligned     256  avgt   30   31,272 ? 0,184  ns/op
```

**Comment by minborg** [comment]
All sizes:


image


Smaller sizes:

image

**Review by mcimadamore — APPROVED** [review]

**Comment by uschindler** [comment]
It would also be good to have a comparison to ByteBuffer (and with LongBuffer views on top).

File: test/micro/org/openjdk/bench/java/lang/foreign/MemorySegmentVsBits.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        b[off + 6] = (byte) (val >>> 8);
+        b[off + 5] = (byte) (val >>> 16);
+        b[off + 4] = (byte) (val >>> 24);
+        b[off + 3] = (byte) (val >>> 32);
+        b[off + 2] = (byte) (val >>> 40);
+        b[off + 1] = (byte) (val >>> 48);
+        b[off] = (byte) (val >>> 56);
+    }
+
+    static long getLong(byte[] b, int off) {
```
Comment (uschindler) [inline test/micro/org/openjdk/bench/java/lang/foreign/MemorySegmentVsBits.java:131]: this method is unused.

**Comment by uschindler** [comment]
Side note:

If you intend to just replace `java.nio.Bits` in JDK, there is another way to do this also without even using Panama. We use the following VarHandles now since long time everywhere in Lucene to put/get long/int/shorts from/to `byte[]`: https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/util/BitUtil.java#L32-L116

This may be cheaper than creating a MemorySegment view on top of java arrays.

**Comment by minborg** [comment]
Thanks @uschindler . I have created another PR that relates to your comments here https://github.com/openjdk/panama-foreign/pull/762

**Pull request closed without merging**